UDN: L2: Add missing remote LRPs

tssurya opened the pull request:
#### What this PR does and why is it needed

In the UDN enhancement https://github.com/openshift/enhancements/blob/master/enhancements/network/user-defined-network-segmentation.md#pod-egress we had discussed about adding remote ports on L2 switch for remote node's GR ports so that communication between GR's can work properly. 

We somehow missed adding this with the L2 main topology.

Also note that we need another fix where I will start changing snat-lrp=joinsubnetIP which will end up reverting the 1st commit in this PR but I'd like to keep that separate from this issue - will have a followup PR for that.

#### Which issue(s) this PR fixes
This fixes that since its required for external->services work on L2.

#### Special notes for reviewers
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->

#### How to verify it
e2e tests are present in services section that supports this

#### Details to documentation updates
<!--
Did you include good docs that explain to our end users/developers/contributors
how your code works?
-->


#### Description for the changelog
`Adds remote LSP port for remote node's GR LRP on L2 UDPNs`

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: TBD
-->
```release-note
NONE
```

tssurya commented:
hmm locally it passes at v4 level and only fails v6, its weird this lane is failing at v4 itself!

qinqon commented:
> hmm locally it passes at v4 level and only fails v6, its weird this lane is failing at v4 itself!

Without he tunnel key collision fix, the L2 test can be failing.

tssurya commented:
https://github.com/ovn-org/ovn-kubernetes/actions/runs/11340936533/job/31539400163?pr=4777
we have a race condition!!! yayy

tssurya commented:
> > hmm locally it passes at v4 level and only fails v6, its weird this lane is failing at v4 itself!
> 
> Without he tunnel key collision fix, the L2 test can be failing.

but I use keys from backside hehe while pods use it from front side, so idallocator doesn't give out 30657 or whatever is the highest id to the pod but but

its a race!! :) I don't see this happening always sooo maybe you are right here :) I will go with fixing this properly as discussed first and see

tssurya commented on go-controller/pkg/ovn/secondary_layer2_network_controller.go:
```diff
@@ -552,6 +556,73 @@ func (oc *SecondaryLayer2NetworkController) addUpdateLocalNodeEvent(node *corev1
+		return fmt.Errorf("failed to get node id for node - %s", node.Name)
+	}
+	logicalSwitchPort.Options = map[string]string{
+		"requested-tnl-key": strconv.Itoa(types.MaxLogicalPortTunnelKey - nodeID),
```
spoke to @jcaamano / @qinqon last week about this.
the easiest way we decided is to use the same pod tunnelID allocator for l2 networks from CM for also the nodes and then annotate the node with the value and use that from here

tssurya commented on go-controller/pkg/clustermanager/node/node_allocator.go:
```diff
@@ -413,8 +438,8 @@ func (na *NodeAllocator) Cleanup() error {
 
 		hostSubnetsMap := map[string][]*net.IPNet{networkName: nil}
-		// passing util.InvalidNetworkID deletes the network id annotation for the network.
-		err = na.updateNodeNetworkAnnotationsWithRetry(node.Name, hostSubnetsMap, util.InvalidNetworkID, nil)
```
oopsy this should have been in commit3 but got sneaked into here..not a biggie though

tssurya commented:
not sure why Live Migration lanes are red?
```
Summarizing 2 Failures:
  [FAIL] Kubevirt Virtual Machines with default pod network when live migration [It] with pre-copy succeeds, should keep connectivity
  /opt/hostedtoolcache/go/1.22.0/x64/src/runtime/asm_amd64.s:1695
  [FAIL] Kubevirt Virtual Machines with default pod network when live migration [It] with pre-copy fails, should keep connectivity
  /opt/hostedtoolcache/go/1.22.0/x64/src/runtime/asm_amd64.s:1695
```
@qinqon PTAL :) adding remote ports was not supposed to break anything right ? :P

qinqon commented:
> not sure why Live Migration lanes are red? https://github.com/ovn-org/ovn-kubernetes/actions/runs/11481726190/job/31954822805?pr=4777 https://github.com/ovn-org/ovn-kubernetes/actions/runs/11481726190/job/31954823187?pr=4777
> 
> ```
> Summarizing 2 Failures:
>   [FAIL] Kubevirt Virtual Machines with default pod network when live migration [It] with pre-copy succeeds, should keep connectivity
>   /opt/hostedtoolcache/go/1.22.0/x64/src/runtime/asm_amd64.s:1695
>   [FAIL] Kubevirt Virtual Machines with default pod network when live migration [It] with pre-copy fails, should keep connectivity
>   /opt/hostedtoolcache/go/1.22.0/x64/src/runtime/asm_amd64.s:1695
> ```
> 
> @qinqon PTAL :) adding remote ports was not supposed to break anything right ? :P

Wow that's the L3 live migration, the remote ports are not even there.

tssurya commented:
> 
> Wow that's the L3 live migration, the remote ports are not even there.

oh then maybe its not related at all! I will retest and see if its a flake or not

tssurya commented:
seems like live migration is flaking on other PRs as well: https://github.com/ovn-org/ovn-kubernetes/pull/4612#issuecomment-2431151104 cc @qinqon seems unrelated to this PR specifically but a headsup

trozet commented on go-controller/pkg/ovn/secondary_layer2_network_controller.go:
```diff
@@ -556,6 +560,74 @@ func (oc *SecondaryLayer2NetworkController) addUpdateLocalNodeEvent(node *corev1
+	tunnelID, err := util.ParseUDNLayer2NodeGRLRPTunnelIDs(node, oc.GetNetworkName())
+	if tunnelID == util.InvalidID {
+		// Don't consider this node as cluster-manager has not allocated node id yet.
+		return fmt.Errorf("failed to fetch tunnelID annotation from the node %s for network %s, err: %v",
```
anytime we have an error where it should be retried, but its expected, we should suppress it. Can you make this a suppressed type error? Maybe the same thing for the join subnet if cluster manager is suppsoed to assign that as well?

tssurya replied:
yeah great call out!

tssurya replied:
done!

tssurya commented on go-controller/pkg/util/node_annotations.go:
```diff
@@ -528,6 +537,34 @@ func ParseNodeManagementPortMACAddresses(node *kapi.Node, netName string) (net.H
+// ParseUDNLayer2NodeGRLRPTunnelIDs parses the 'ovnUDNLayer2NodeGRLRPTunnelIDs' annotation
+// for the specified network in 'netName' and returns the tunnelID.
+func ParseUDNLayer2NodeGRLRPTunnelIDs(node *kapi.Node, netName string) (int, error) {
+	networkIDsMap, err := parseNetworkMapAnnotation(node.Annotations, ovnUDNLayer2NodeGRLRPTunnelIDs)
```
ah oopsy :D

tssurya replied:
fixed.

tssurya commented on go-controller/pkg/util/node_annotations.go:
```diff
@@ -1328,14 +1328,15 @@ func ParseNetworkIDAnnotation(node *kapi.Node, netName string) (int, error) {
+// with the provided ID in 'annotationName”s value.  If 'ID' is InvalidNetworkID (-1)
+// it deletes the annotationName annotation from the map.
+// It is currently used for ovnNetworkIDs annotation updates
+func updateNetworkAnnotation(annotations map[string]string, netName string, ID int, annotationName string) error {
```
ack let me squash those two commits so that its less confusing

qinqon commented on go-controller/pkg/ovn/secondary_layer2_network_controller.go:
```diff
@@ -556,6 +560,74 @@ func (oc *SecondaryLayer2NetworkController) addUpdateLocalNodeEvent(node *corev1
+		remoteGRPortNetworks = append(remoteGRPortNetworks, ip.String())
+	}
+
+	remotePortAddr := remoteGRPortMac.String()
+	for _, tsNetwork := range remoteGRPortNetworks {
+		remotePortAddr = remotePortAddr + " " + tsNetwork
+	}
```
Is it worth it to use strings.Join to remove the loop here ? 
```golang
remotePortAddr :=  remoteGRPortMac.String() +  " " + strings.Join(remoteGRPortNetworks, " ")
```
Don't know if this improve the readability.

tssurya replied:
sure  I don't mind doing that

qinqon commented on go-controller/pkg/ovn/secondary_layer2_network_controller.go:
```diff
@@ -556,6 +560,74 @@ func (oc *SecondaryLayer2NetworkController) addUpdateLocalNodeEvent(node *corev1
+	logicalSwitchPort.ExternalIDs = map[string]string{
+		types.NetworkExternalID:  oc.GetNetworkName(),
+		types.TopologyExternalID: oc.TopologyType(),
+		"node":                   node.Name,
```
Let's create another const for it at types.go

```golang
NodeExternalID = OvnK8sPrefix + "/" + "node"
```

tssurya replied:
ideally all this should be going into DBIndexes.. but until we reach that cleanup phase let's do what you suggested....

tssurya replied:
done.

qinqon commented on go-controller/pkg/ovn/secondary_layer2_network_controller.go:
```diff
@@ -556,6 +560,74 @@ func (oc *SecondaryLayer2NetworkController) addUpdateLocalNodeEvent(node *corev1
+		"node":                   node.Name,
+	}
+	tunnelID, err := util.ParseUDNLayer2NodeGRLRPTunnelIDs(node, oc.GetNetworkName())
+	if tunnelID == util.InvalidID {
```
feels weird to not have the typical `if err != nil`, maybe is better to have `if err != nil` and `if tunnelID == util.InvalidID` returning different errors.

tssurya replied:
yup I felt the same as I read my own code :) thanks for pointing this out, fixed it.

qinqon reviewed (changes requested):
Nothing big, I am going to use this PR to test live migration manually, since we don't test egress at live migration.

tssurya commented on go-controller/pkg/util/node_annotations.go:
```diff
@@ -1328,14 +1328,15 @@ func ParseNetworkIDAnnotation(node *kapi.Node, netName string) (int, error) {
+// with the provided ID in 'annotationName”s value.  If 'ID' is InvalidNetworkID (-1)
+// it deletes the annotationName annotation from the map.
+// It is currently used for ovnNetworkIDs annotation updates
+func updateNetworkAnnotation(annotations map[string]string, netName string, ID int, annotationName string) error {
```
yeah I can make it "id", the reason why I changed this was I was about to add another "invalidTunnelID" with -1 as the value, we already have invalidNodeID and invalidNetworkID, if we keep adding more -1's I felt its better to consolidate them if there is commonality...

qinqon commented:
@tssurya after starting the PR up at kind.sh I found that we may need to also change the "router" type LSP, I think remote LSPs should have the same name as local LSPs 
```
   port stor-l2.network_ovn_layer2_switch_ovn-worker2
        type: remote
        addresses: ["0a:58:64:41:00:04 100.65.0.4/16 fd99::4/64"]
    port stor-l2.network_ovn_layer2_switch <---- this should be "stor-l2.network_ovn_layer2_switch_ovn-worker"
        type: router
        router-port: rtos-l2.network_ovn_layer2_switch
```

tssurya commented:
some local port options are missing... fixing them and also addressing review comments

qinqon commented on test/e2e/network_segmentation_services.go:
```diff
@@ -159,14 +158,12 @@ var _ = Describe("Network Segmentation: services", func() {
-						checkConnectionToNodePort(f, udnClientPod, udnService, &nodes.Items[1], "other node", udnServerPod.Name)
-						checkConnectionToNodePort(f, udnClientPod, udnService, &nodes.Items[2], "other node", udnServerPod.Name)
-					}
+					checkConnectionToNodePort(f, udnClientPod, udnService, &nodes.Items[1], "other node", udnServerPod.Name)
+					checkConnectionToNodePort(f, udnClientPod, udnService, &nodes.Items[2], "other node", udnServerPod.Name)
```
This is not going to exercise the remote port since it's using the remote pod's LSP.

qinqon replied:
We can ignore this, since e2e is testing it within a pod, so the remote LSPs are exercised.

qinqon replied:
The remote LSP are not exercised even if the client is at the pod since the OVN LBs convert the nodeport into pod's IP, so traffic goes over the pod's remote LSP.

tssurya replied:
no there is a specific test that exercises the remote port LRP, I tested it using that test :)

qinqon replied:
Righ right, my bad I have just take into account that the pod has to answer back using the remote node IP, and for that we need the remote ports... we can totally disregard my comment

qinqon replied:
But then how can be possible that test are passing without local ports having the tunnel key ?

tssurya replied:
exactly.. hehe I think there is a slight race somewhere .... let's just fix the code to be correct and see

tssurya commented on go-controller/pkg/util/node_annotations.go:
```diff
@@ -143,7 +143,16 @@ const (
+	//		"l2-network-a":"5",
+	//		"l2-network-b":"10"}
+	// }",
+	ovnUDNLayer2NodeGRLRPTunnelIDs = "k8s.ovn.org/udn-layer2-node-gateway-router-lrp-tunnel-ids"
```
hmm does nodeTracker do something for this? I thought its only used in services area..

trozet replied:
it does, but it fires on every node change. We have n number of nodeTrackers for n number of UDNs. @dceara is going to abstract it to only have a single node tracker to improve perf

tssurya commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -172,10 +174,47 @@ func (ncc *networkClusterController) init() error {
+				tunnelID, err := util.ParseUDNLayer2NodeGRLRPTunnelIDs(node, ncc.GetNetworkName())
+				if err != nil {
+					if util.IsAnnotationNotSetError(err) {
+						fmt.Printf("tunnelID annotation does not exist for the node %s for network %s, err: %v; we need to allocate it...",
```
I meant Infof ouchy

tssurya commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -172,10 +174,47 @@ func (ncc *networkClusterController) init() error {
+		if err != nil {
+			return err
+		}
+		if util.IsNetworkSegmentationSupportEnabled() && ncc.IsPrimaryNetwork() {
```
no no we need the first part of this code also for secondary networks.. secondary L2 networks need the idallocator to give out the tunnelIDs..

tssurya commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -172,10 +174,47 @@ func (ncc *networkClusterController) init() error {
 		return err
 	}
 
+	if util.DoesNetworkRequireTunnelIDs(ncc.NetInfo) {
```
no no we need the first part of this code also for secondary networks.. secondary L2 networks need the idallocator to give out the tunnelIDs..so all the things till the nodeLRPs tunnelID allocation is common for both primary and secondary... the node LRP reserve is a primary only thing

tssurya commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -172,10 +174,47 @@ func (ncc *networkClusterController) init() error {
+	if util.DoesNetworkRequireTunnelIDs(ncc.NetInfo) {
+		ncc.tunnelIDAllocator, err = id.NewIDAllocator(ncc.GetNetworkName(), types.MaxLogicalPortTunnelKey)
+		if err != nil {
+			return fmt.Errorf("failed to create new id allocator for network %s, err: %v", ncc.GetNetworkName(), err)
```
done

tssurya commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -172,10 +174,47 @@ func (ncc *networkClusterController) init() error {
+			return fmt.Errorf("failed to create new id allocator for network %s, err: %v", ncc.GetNetworkName(), err)
+		}
+		// Reserve the id 0. We don't want to assign this id to any of the pods or nodes.
+		err = ncc.tunnelIDAllocator.ReserveID("zero", 0)
+		if err != nil {
+			return err
+		}
```
this was copy-paste from previous location :) but can change this..

tssurya commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -172,10 +174,47 @@ func (ncc *networkClusterController) init() error {
+							node.Name, ncc.GetNetworkName(), err)
+					}
+				}
+				if tunnelID != util.InvalidID {
```
right so if the opposite happens we would have caught it in the "err!=nil" i.e err!= nil is same as invalidID as tunnelID.. so should I remove this explicit check?

qinqon commented on go-controller/pkg/clustermanager/node/node_allocator.go:
```diff
@@ -276,10 +279,26 @@ func (na *NodeAllocator) syncNodeNetworkAnnotations(node *corev1.Node) error {
 		}
 	}
+	newTunnelID := 0
+	if util.IsNetworkSegmentationSupportEnabled() && na.netInfo.IsPrimaryNetwork() && util.DoesNetworkRequireTunnelIDs(na.netInfo) {
```
Update: not needed.

tssurya replied:
ditto as my previous response...

tssurya commented on go-controller/pkg/clustermanager/node/node_allocator.go:
```diff
@@ -276,10 +279,26 @@ func (na *NodeAllocator) syncNodeNetworkAnnotations(node *corev1.Node) error {
+		} else {
+			err = na.idAllocator.ReserveID(networkName+"_"+node.Name, existingTunnelID)
+		}
+		if err != nil {
```
ack I had it for if and else...hence thought of making it common but it can be confusing to do that.. I can move this to if and else pieces...

tssurya commented on go-controller/pkg/clustermanager/node/node_allocator.go:
```diff
@@ -276,10 +279,26 @@ func (na *NodeAllocator) syncNodeNetworkAnnotations(node *corev1.Node) error {
-	if len(updatedSubnetsMap) > 0 || na.networkID != networkID || len(allocatedJoinSubnets) > 0 {
-		err = na.updateNodeNetworkAnnotationsWithRetry(node.Name, updatedSubnetsMap, na.networkID, allocatedJoinSubnets)
+	if len(updatedSubnetsMap) > 0 || na.networkID != networkID || len(allocatedJoinSubnets) > 0 || newTunnelID != 0 {
+		err = na.updateNodeNetworkAnnotationsWithRetry(node.Name, updatedSubnetsMap, na.networkID, newTunnelID, allocatedJoinSubnets)
```
good catch! I want to say while having a shower this scenario hit me.. that I forgot to do this..fixing this

tssurya commented on go-controller/pkg/clustermanager/node/node_allocator.go:
```diff
@@ -378,12 +397,18 @@ func (na *NodeAllocator) updateNodeNetworkAnnotationsWithRetry(nodeName string,
 			return fmt.Errorf("failed to update node %q network id annotation %d for network %s",
 				node.Name, networkId, networkName)
 		}
+		if tunnelID != 0 {
```
ack did a NoID constant for clarity

tssurya commented on go-controller/pkg/clustermanager/node/node_allocator.go:
```diff
@@ -378,12 +397,18 @@ func (na *NodeAllocator) updateNodeNetworkAnnotationsWithRetry(nodeName string,
+		if tunnelID != 0 {
+			cnode.Annotations, err = util.UpdateUDNLayer2NodeGRLRPTunnelIDs(cnode.Annotations, networkName, tunnelID)
+			if err != nil {
+				return fmt.Errorf("failed to update node %q tunnel id annotation %d for network %s",
```
done

tssurya commented on go-controller/pkg/ovn/secondary_layer2_network_controller.go:
```diff
@@ -556,6 +560,74 @@ func (oc *SecondaryLayer2NetworkController) addUpdateLocalNodeEvent(node *corev1
+
+func (oc *SecondaryLayer2NetworkController) addPortForRemoteNodeGR(node *corev1.Node) error {
+	nodeJoinSubnetIPs, err := util.ParseNodeGatewayRouterJoinAddrs(node, oc.GetNetworkName())
+	if err != nil || len(nodeJoinSubnetIPs) == 0 {
+		return fmt.Errorf("failed to get the node %s join subnet IPs : %w", node.Name, err)
+	}
```
good callout, fixed

tssurya commented on go-controller/pkg/ovn/secondary_layer2_network_controller.go:
```diff
@@ -556,6 +560,74 @@ func (oc *SecondaryLayer2NetworkController) addUpdateLocalNodeEvent(node *corev1
+		return fmt.Errorf("failed to get the node %s join subnet IPs : %w", node.Name, err)
+	}
+
+	remoteGRPortMac := util.IPAddrToHWAddr(nodeJoinSubnetIPs[0].IP)
```
from what I could tell this same util is used in other places of the code...specially on the gateway side:

```
gwLRPMAC := util.IPAddrToHWAddr(gwLRPIPs[0])
	gwLRPNetworks := []string{}
	for _, gwLRPIfAddr := range gwLRPIfAddrs {
		gwLRPNetworks = append(gwLRPNetworks, gwLRPIfAddr.String())
	}
```
hence stuck to the same method that is being used to create the local LRP.. I agree we need a more formal util to avoid mistakes but that can be a code cleanup FUP

tssurya commented on go-controller/pkg/util/node_annotations.go:
```diff
@@ -143,8 +143,8 @@ const (
-	// invalidNetworkID signifies its an invalid network id
-	InvalidNetworkID = -1
+	// InvalidID signifies its an invalid network id
+	InvalidID = -1
```
given we have been using invalidNetworkID everywhere... just assumed to continue that slang...

tssurya commented:
> some local port options are missing... fixing them and also addressing review comments

this should be fixed now.. let me spin up a local one based on latest changes

tssurya commented:
I think it looks good now:
```
                    k8s.ovn.org/udn-layer2-node-gateway-router-lrp-tunnel-ids: {"ns6.layer2-crd-network":"1"}                                                                
```
localport:
```
sh-5.2# ovn-nbctl list logical-switch-port stor-ns6.layer2.crd.network_ovn_layer2_switch
_uuid               : 1121b195-cdad-4da4-b3ef-5c969f1f5022
addresses           : [router]
dhcpv4_options      : []
dhcpv6_options      : []
dynamic_addresses   : []
enabled             : []
external_ids        : {"k8s.ovn.org/network"=ns6.layer2-crd-network, "k8s.ovn.org/topology"=layer2}
ha_chassis_group    : []
mirror_rules        : []
name                : stor-ns6.layer2.crd.network_ovn_layer2_switch
options             : {requested-tnl-key="1", router-port=rtos-ns6.layer2.crd.network_ovn_layer2_switch}
parent_name         : []
port_security       : []
tag                 : []
tag_request         : []
type                : router
up                  : true
```
remote-port
```
sh-5.2# ovn-nbctl list logical-switch-port stor-ns6.layer2.crd.network_ovn_layer2_switch_ovn-control-plane                                                                   
_uuid               : c73c451e-c1c0-49be-8e54-67e2fc1ee475
addresses           : ["0a:58:64:41:00:02 100.65.0.2/16"]
dhcpv4_options      : []
dhcpv6_options      : []
dynamic_addresses   : []
enabled             : []
external_ids        : {"k8s.ovn.org/network"=ns6.layer2-crd-network, "k8s.ovn.org/node"=ovn-control-plane, "k8s.ovn.org/topology"=layer2}
ha_chassis_group    : []
mirror_rules        : []
name                : stor-ns6.layer2.crd.network_ovn_layer2_switch_ovn-control-plane
options             : {requested-chassis=ovn-control-plane, requested-tnl-key="1"}
parent_name         : []
port_security       : []
tag                 : []
tag_request         : []
type                : remote
up                  : true
```

qinqon commented:
> requested-chassis

I think local port is missing the requested-chassis

dceara commented:
> > > requested-chassis
> > 
> > 
> > I think local port is missing the requested-chassis
> 
> no local port doesn't need it check the transit switch ports on l3.. I don't see them there.. can you confirm @dceara
> 

You're right, there was some confusion here, no need to set it for local chassis.

> > Also would be good to rename local port so it follows the same name like remote ports (adding the node name to it)
> 
> I left it intentionally like that :P I really wanted to know from "show command" which ones the local one...

This sounds ok to me.

tssurya commented:
tracking shard-failure: https://github.com/ovn-org/ovn-kubernetes/issues/4806 its unrelated to this change
